### DISCUSSION TITLE
Support for two VEdirect

### DISCUSSION THREAD

**Discussion by sicanins** [hoylabs/OpenDTU-OnBattery#512] [Show and tell]
FYI I added support for two VEdirect in my fork. If anybody is interested, I could spend some time to prepare a pull request.

[https://github.com/sicanins/OpenDTU-OnBattery](https://github.com/sicanins/OpenDTU-OnBattery)

@helgeerbe

**Comment by Bitfluesterer** [comment]
That's great. Thank you very much. I have 2 Victron 100/20 charge controllers in use and have been looking for a solution to integrate them. In the future I would like to expand my solar system and add a third charge controller e.g. 150/35. How could this be integrated? Is there a possibility?

**Comment by schlimmchen** [comment]
I see what you did there and I understand that it works for you. However, I have strong objections concerning your approach:

* Creating multiple static instances of MPPT charge controllers is inefficient for everybody that does not use multiple charge controllers or even none at all.
* Looping over all these instances in so many different places makes no sense to me. Not only is it code duplication all over the place, it's mixing concerns and missing separation. Please have a look at https://github.com/helgeerbe/OpenDTU-OnBattery/blob/development/src/VictronMppt.cpp
* You did not solve the issue where the ESP32 will run out of hardware UARTs. Moreover, you created a conflict with the JK BMS instance, which uses HW UART with index 2. Your second MPPT instance will conflict with that. A fourth one (index 3) will simply fail to initialize, since there is not fourth hardware UART.

**Comment by Napizaki** [comment]
Hello everyone,
I am currently fully satisfied with the OpenDTU-onBattery solution - many thanks to everyone involved!
I am planning to expand my current system (2 x Panel, 1 Victron 100/20) with 2 panels and therefore also a second Victron 100/20. What would be the best solution at the moment for integrating the second Victron into the system so that I can continue to use the DPL including solar passthrough?

Also connected to the DTU:
- Shelly 3em
- Battery: PT us2000c
- Hoymiles HM-600

**Comment by sicanins** [comment]
Hi @schlimmchen,

let me try to answer to your points with my point of view:

1) as far as I understand the object oriented programming, its exactly the idea to have multiple instances of the same object? Where is the advantage of adding the loops into the object and only creating one instance?
2) same answer on the looping. Its only two instances ever, so I dont see the problem. If you find this concerning, then why is the code looping so many other objects every time even they are not needed?
It could be improved by setting the VICTRON_COUNT correctly, but I was not able to identify the include file in which I could do that. Seems there is not one single config file thats included by all other files later?
I am not clear about the "separation" part. Where is the code missing mutexes? Help is appreciated here, I did not see clear.
3) Thats a major issue obivously. Two approaches:
a) swapping the existing UART between different pins: checked the theory behind this. The problem is, HEX communication is needed to query the information actively. The driver for the MPPTs is then much more difficult as it needs to know about different models and version of the Victrons, and needs to be updated when new models are coming out. Thats a major drawback I think also for testing in the future. Too many variables. 
Waiting on different pins for the incoming data makes the system very slow, as my code now needs the info from the MPPT to run the zero feed in faster.
b) external UARTs. Can be done by SPI UARTs or coprocessors (ex. small ARMs). Did this before on several projects and it is quite difficult, just think of the bootloader for the coprocessor. Also a lot of variables.

Thats why my pragmatic conclusion was and is to use the built in UARTs and limit the combinations of what can be used together. In other words, I feel its essentially a limit of the ESP32. 

I am happy to discuss this :)

Thanks
